src/handlers/web3.ts

```diff
@@ -0,0 +1,681 @@
+import { getAddress } from '@ethersproject/address';
+import { BigNumber } from '@ethersproject/bignumber';
+import { isHexString as isEthersHexString } from '@ethersproject/bytes';
+import { Contract } from '@ethersproject/contracts';
+import { isValidMnemonic as ethersIsValidMnemonic } from '@ethersproject/hdnode';
+import {
+  Block,
+  JsonRpcProvider,
+  TransactionRequest,
+  TransactionResponse,
+} from '@ethersproject/providers';
+import { parseEther } from '@ethersproject/units';
+import UnstoppableResolution from '@unstoppabledomains/resolution';
+import { get, startsWith } from 'lodash';
+import { RainbowConfig } from '../model/config';
+import {
+  AssetType,
+  NewTransaction,
+  ParsedAddressAsset,
+} from '@rainbow-me/entities';
+import { isNativeAsset } from '@rainbow-me/handlers/assets';
+import { Network } from '@rainbow-me/helpers/networkTypes';
+import {
+  ARBITRUM_ETH_ADDRESS,
+  ETH_ADDRESS,
+  ethUnits,
+  MATIC_POLYGON_ADDRESS,
+  OPTIMISM_ETH_ADDRESS,
+  smartContractMethods,
+} from '@rainbow-me/references';
+import {
+  addBuffer,
+  convertAmountToRawAmount,
+  convertStringToHex,
+  fraction,
+  greaterThan,
+  handleSignificantDecimals,
+  multiply,
+} from '@rainbow-me/utilities';
+import { ethereumUtils } from '@rainbow-me/utils';
+import logger from 'logger';
+
+export const networkProviders: { [network in Network]?: JsonRpcProvider } = {};
+
+const rpcEndpoints: { [network in Network]?: string } = {};
+
+/**
+ * The format of transaction details returned by functions such as `getTxDetails`.
+ */
+interface TransactionDetailsReturned {
+  data?: TransactionRequest['data'];
+  to?: TransactionRequest['to'];
+  value?: TransactionRequest['value'];
+  from?: TransactionRequest['from'];
+  gasLimit?: string;
+  gasPrice?: string;
+}
+
+/**
+ * A type with the same keys as `NewTransaction` but without nullable types.
+ * This is useful for functions that assume that certain fields are not set
+ * to null on a `NewTransaction`.
+ */
+type NewTransactionNonNullable = {
+  [key in keyof NewTransaction]: NonNullable<NewTransaction[key]>;
+};
+
+/**
+ * @desc Configures `rpcEndpoints` based on a given `RainbowConfig`.
+ * @param config The `RainbowConfig` to use.
+ */
+export const setRpcEndpoints = (config: RainbowConfig) => {
+  rpcEndpoints[Network.mainnet] = config.ethereum_mainnet_rpc;
+  rpcEndpoints[Network.ropsten] = config.ethereum_ropsten_rpc;
+  rpcEndpoints[Network.kovan] = config.ethereum_kovan_rpc;
+  rpcEndpoints[Network.rinkeby] = config.ethereum_rinkeby_rpc;
+  rpcEndpoints[Network.optimism] = config.optimism_mainnet_rpc;
+  rpcEndpoints[Network.arbitrum] = config.arbitrum_mainnet_rpc;
+  rpcEndpoints[Network.polygon] = config.polygon_mainnet_rpc;
+};
+
+/**
+ * @desc web3 http instance
+ */
+export let web3Provider: JsonRpcProvider = (null as unknown) as JsonRpcProvider;
```

Comment:
This should be typed as `JsonRpcProvider | null`, however, there are many places throughout the codebase that assume `web3Provider` is non-null. Either we can cast it to `JsonRpcProvider` here, update all of those instances to use a non-null assertion, or update them to fail/return undefined if `web3Provider` is not set.

Reply:
Added RNBW-1811 Linear ticket for this.